### ISSUE TITLE
"bs-platform could not be found" when opening sublime text

### ISSUE THREAD

**Issue by mnieber** [jaredly/reason-language-server#210]
Hi, 

I'm getting a dialog window saying "bs-platform could not be found" when opening sublime text. I'm using the option to enable LSP just for the current project. I've pasted the debug.log below. 

The language server should work completely independently from my project's node installation, right? In my case, I run everything inside a docker container, and I'm only using the host system to edit the source, which means that on the host computer there is no node_modules at all.

I do see that the language server has created a new `node_modules` folder in my `src` folder, but this folder only contains the `.lsp/debug.log` file.

Thanks in advance,
Maarten

    Hello from /home/maarten/bin/reason-language-server/reason-language-server.exe
    Previous log location: /tmp/lsp.log
    Sending notification {"jsonrpc": "2.0", "method": "client/registerCapability", "params": {"registrations": [{"id": "watching", "method": "workspace/didChangeWatchedFiles", "registerOptions": {"watchers": [{"globPattern": "**/bsconfig.json", "globPattern": "**/.merlin"}]}}]}}
    Sending response {"id": 1, "jsonrpc": "2.0", "result": {"capabilities": {"textDocumentSync": 1, "hoverProvider": true, "completionProvider": {"resolveProvider": true, "triggerCharacters": ["."]}, "signatureHelpProvider": {"triggerCharacters": ["("]}, "definitionProvider": true, "typeDefinitionProvider": true, "referencesProvider": true, "documentSymbolProvider": true, "codeActionProvider": true, "executeCommandProvider": {"commands": ["reason-language-server.add_to_interface_inner"]}, "codeLensProvider": {"resolveProvider": true}, "documentHighlightProvider": true, "documentRangeFormattingProvider": true, "documentFormattingProvider": true, "documentFormattingProvider": true, "renameProvider": true}}}
    Read message 
    {"jsonrpc": "2.0", "method": "textDocument/didOpen", "params": {"textDocument": {"version": 0, "text": "let str = ReasonReact.string;\n\nmodule Decoding = {\n  open Move;\n  let move = move =>\n    {\n      id: Json.Decode.field(\"id\", Json.Decode.string, move),\n      name: Json.Decode.field(\"name\", Json.Decode.string, move),\n    };\n  let moves = json : list(Move.t) => Json.Decode.list(move, json);\n};\n\nlet fetchMoves = () =>\n  Js.Promise.(\n    Fetch.fetch(\"http://localhost:8000/moves\")\n    |> then_(Fetch.Response.json)\n    |> then_(json => json |> decodeObject |> Js.dict.get(_, \"moves\") |> resolve)\n    |> then_(json =>\n         json |> Decoding.moves |> (moves => {Js.log2(\"CCC:\", moves); Some(moves)} |> resolve)\n       )\n    |> catch(_err => resolve(None))\n  );\n\nfetchMoves();\nReactDOMRe.renderToElementWithId(<h1>{str(\"HA\")}</h1>, \"root\");\n", "languageId": "reason", "uri": "file:///home/maarten/projects/linsci/src/website/src/Main.re"}}}
    bs-platform could not be found
    Sending notification {"jsonrpc": "2.0", "method": "window/showMessage", "params": {"type": 1, "message": "bs-platform could not be found"}}
    bs-platform could not be found

**Comment by jaredly** [comment]
ah that's fascinating. Yeah, it would be super cool to be able to support this. we've run into a similar issue with windows' "linux mode" (WSL) where the host architecture (that vscode was running in) was different from the project architecture (where node_modules was installed).
I'm impressed you were able to get sublime to run rsl in the docker container! did you change the command to `docker run mycontainer /path/to/rsl.exe` or something?

It sounds like what you want is a way to say "when you see a subdirectory of /path/to/X from sublime, translate it to a subdirectory of /path/to/Y before going to the filesystem". Is that right? I don't think it would be too hard to add a setting for that.

**Comment by mnieber** [comment]
Hi Jared, yes, my command setting is now: 

    "command": [
        "docker", 
        "run", 
        "--interactive", "--rm", 
        "--volume=/home/maarten/projects/linsci/src:/srv/linsci/src", 
        "--volumes-from=dc_linsci_packages", 
        "--workdir=/srv/linsci/src/website", 
        "linsci:base", 
        "/srv/linsci/src/website/node_modules/.bin/reason-language-server.exe"
    ],

Note that by using `volumes-from=dc_linsci_packages`, I'm dropping in a `node_modules` directory into my source directory at location `/srv/linsci/src/website/node_modules`.

Your proposal of translating paths is one way to do it. Another way might be to have a setting that directly points to the right `node_modules` folder and right `root path` (so that you don't need to find them based on the current filename). In my opinion this option (if it works) would be a bit better because the intention of these settings would be clearer, and it would support cases where the location of `node_modules` is unrelated to the source filename.

Another small issue is that you would ideally choose these package-specific settings (including the "command" setting) per project. Perhaps a solution is to allow the user to store settings in a file inside the project tree (similar to `bsconfig.json`) instead of in `LSP.sublime-settings` (or would this run into the same problem of not being able to find the file?)

**Comment by jaredly** [comment]
Yeah, I like that idea a lot. Looking at your docker config, I think passing in something like `/home/maarten/projects/linsci/src:/srv/linsci/src` would be all we'd need to be able to determine where source files are.

And we definitely want the configuration to be per-project. I'm not sure how sublime's per-project LSP config works, but in vscode the configuration goes in `<project-root>/.vscode/settings.json`, and in vim you can put it in `<project-root>/.vim/settings.json` ([see readme](https://github.com/jaredly/reason-language-server#configuration)). Maybe sublime has something similar?

**Comment by mnieber** [comment]
As far as I know, there are no per-project package settings at all in sublime. 

I should say that I've parked my ReasonReact efforts for the moment, as I did not find a good Redux alternative for managing a global app state, but I could still help with making this docker setup work.
